---
 configure.ac |   21 +++------------------
 1 file changed, 3 insertions(+), 18 deletions(-)

--- systemd-204.orig/configure.ac
+++ systemd-204/configure.ac
@@ -450,25 +450,10 @@ AC_ARG_ENABLE([gcrypt],
                 esac],
                 [have_gcrypt=auto])
 
-if test "x${have_gcrypt}" != xno ; then
-        AM_PATH_LIBGCRYPT(
-                [1.4.5],
-                [have_gcrypt=yes],
-                [if test "x$have_gcrypt" = xyes ; then
-                        AC_MSG_ERROR([*** GCRYPT headers not found.])
-                fi])
+have_gcrypt=no
+GCRYPT_LIBS=
+GCRYPT_CFLAGS=
 
-        if test "x$have_gcrypt" = xyes ; then
-                GCRYPT_LIBS="$LIBGCRYPT_LIBS"
-                GCRYPT_CFLAGS="$LIBGCRYPT_CFLAGS"
-                AC_DEFINE(HAVE_GCRYPT, 1, [GCRYPT available])
-        else
-                have_gcrypt=no
-        fi
-else
-        GCRYPT_LIBS=
-        GCRYPT_CFLAGS=
-fi
 AC_SUBST(GCRYPT_LIBS)
 AC_SUBST(GCRYPT_CFLAGS)
 AM_CONDITIONAL([HAVE_GCRYPT], [test "x$have_gcrypt" != xno])
